### PR TITLE
Cledi tyler shanmugam pull 13

### LINKED ISSUE #13: Enrolling New Chapter (University)
## Problem

The Legacy Project would like host multiple universities on their platform. The concept of a "chapter" does not yet exist in the application; therefore, we will need build the backend to support this feature. Looking ahead, your implementation will allow for the following user flow:

Image

1. A user submits a request to start a new chapter,
2. An admin approves the request,
3. The new chapter will be added to the admin's view.

## Specifications

Your task is to define a `Chapter` schema and a backend API route for users to request a new chapter. 

### Chapter Schema

We don't fully know what fields to include in the `Chapter` schema yet, but at the minimum, it should contain:

1. The university name,
2. References to its members.

You don't have to worry about migrating `User` document to include references to their chapter; we will update that in future sprints.

### API Route

You will define the backend API route in `./src/app/api`. The information required to start a new chapter can be found in this [form](https://docs.google.com/forms/d/e/1FAIpQLScmJkAJhUNtIQSOjyvihAI5w_fEh7soZQ3fDLfCrA5KTbAR7g/viewform). After a request to create a chapter has been submitted, the admin will need to review it; therefore, you will need to figure out how to best persist the information. In addition, we would like you to follow this [style guide](https://github.com/JumboCode/TheLegacyProject2/tree/main/src/app/api/toy-example) when designing the API route.

### Pointers

We anticipate that you will be working with `Prisma`, `Zod`, `Next.js 13`, and `MongoDB` - you can find related documents for these frameworks on Google. If you're still stuck afterwards, please reach out to us! Additionally, we've included some resources that you may find helpful:

1. [Next.js 13 route handlers](https://nextjs.org/docs/app/building-your-application/routing/route-handlers),
2. You will need to interact with the database through Prisma, which you can find in`./src/server/db/client.ts`,
3. You will use `./prisma/schema.prisma` to define the new schemas,
4. If you're defining complicated `Zod` schemas, it will make sense to write some unit tests. Your tests will save us hours of frustration down the line :)

### PR DESCRIPTION
**Pull request by tylerrlin**
# Description

Schema changes to allow for new chapter request and API route to request a new chapter.

## Issues

Resolve #13 
Enrolling New Chapter (University)

## Screenshots

Only changes to backend

## Test

Postman POST requests to localhost:3000/api/chapter-request/

## Possible Downsides

Phone number not validated (just a string input)

## Additional Documentations

Sample POST request:

{
    "firstName": "Tyler",
    "lastName": "Lin",
    "universityEmail": "tlin08@tufts.edu",
    "phoneNumber": "123-123-1234",
    "university": "Tufts",
    "universityAddress": "123 Packard Ave",
    "leadershipExperience": "Some",
    "motivation": "Some",
    "availabilities": "Always",
    "questions": "None",
}

### CODE REVIEW

File: src/app/api/chapter-request/route.schema.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  leadershipExperience String
+  motivation           String
+  availabilities       String
+  questions            String
+*/
+export const ChapterRequest = z.object({
+  firstName: z.string(),
+  lastName: z.string(),
+  universityEmail: z.string().email("This is not a valid email"),
+  phoneNumber: z.string(),
```
Comment (nickbar01234) [inline src/app/api/chapter-request/route.schema.ts:19]: We can enforce that phone numbers are valid. For now, I think it's safe to assume that the numbers are in United States - i.e 10 digits number

File: prisma/schema.prisma [hunk 2]
```diff
@@ -96,4 +96,19 @@ model File {
...
+}
+
+model ChapterRequest {
+  id                   String    @id @default(auto()) @map("_id") @db.ObjectId
+  approved             Approval  @default(PENDING)
+  firstName            String
+  lastName             String
+  universityEmail      String    @unique
+  phoneNumber          String
+  university           String
```
Comment (nickbar01234) [inline prisma/schema.prisma:108]: Should university be unique too? Is there a case when we might have 2 universities of the same name?

File: src/app/api/chapter-request/route.schema.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  university: z.string(),
+  universityAddress: z.string(),
+  leadershipExperience: z.string(),
+  motivation: z.string(),
+  availabilities: z.string(),
+  questions: z.string(),
+});
```
Comment (nickbar01234) [inline src/app/api/chapter-request/route.schema.ts:26]: Before the semi-colon, add `satisfies z.ZodType<Prisma.ChapterRequestCreateInput>`; this ensures that your schema matches the expected database type. `npm run build` will fail if there is a mismatch.

File: src/app/api/chapter-request/route.schema.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+export const ChapterRequest = z.object({
+  firstName: z.string(),
+  lastName: z.string(),
+  universityEmail: z.string().email("This is not a valid email"),
+  phoneNumber: z.string(),
+  university: z.string(),
+  universityAddress: z.string(),
+  leadershipExperience: z.string(),
+  motivation: z.string(),
+  availabilities: z.string(),
```
Comment (nickbar01234) [inline src/app/api/chapter-request/route.schema.ts:24]: I would leave a todo here. Passing a list of time availabilities as a string seems a bit counterintuitive. But we also don't fully know how the frontend might choose to parse the time.

**Review by nickbar01234 — CHANGES_REQUESTED** [review]
Hey team, great working adding the API endpoint to create a new chapter. There's a couple of minor changes I would like to see.

1. Add a constraint for `phoneNumber` length in `ChapterRequest` schema.
2. Add `satisfies z.ZodType<Prisma.ChapterRequestInput>` to `ChapterRequest` schema. This forces your zod object to contain all the fields in the Prisma schema.
3. Create a "client-facing" function so that frontend developers can call your API.
4. Finally, I think the cases for `ChapterRequestResponse` can be condensed into 4 cases: `SUCCESS`, `UNKOWN`, `DUPLICATE_EMAIL`, and `INVALID_FORM`. The last case `INVALID_FORM` happens when `ChapterRequest` fails to parse. We don't need all the cases because `ChapterRequest` schema will also be used by the frontend developers to provide feedback for the user filling out the form. With this schema changes, you will also need to update `route.ts` to reflect the new changes.

File: src/app/api/chapter-request/route.schema.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  questions: z.string(),
+});
+
+export const ChapterRequestResponse = z.discriminatedUnion("code", [
+  z.object({
+    code: z.literal("SUCCESS"),
+    message: z.literal("Chapter request successfully submitted"),
+    data: z.any(),
```
Comment (nickbar01234) [inline src/app/api/chapter-request/route.schema.ts:32]: Should this be `z.any()`? I don't think we need to send back any data from backend when a user requests to start a chapter - what are your thoughts?

File: src/app/api/chapter-request/route.client.ts [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+  if (mock === "SUCCESS") {
+    return ChapterRequestResponse.parse(MOCK_SUCCESS);
+  } else if (mock === "INVALID_FORM") {
+    return ChapterRequestResponse.parse(MOCK_INVALID_FORM);
+  } else if (mock === "DUPLICATE_EMAIL") {
+    return ChapterRequestResponse.parse(MOCK_DUPLICATE_EMAIL);
+  } else if (mock === "UNKNOWN") {
+    return ChapterRequestResponse.parse(MOCK_UNKNOWN);
+  }
+  console.log("here");
```
Comment (nickbar01234) [inline src/app/api/chapter-request/route.client.ts:64]: nit: Remove redundant console

File: src/app/api/chapter-request/route.client.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+  message: "Invalid form submission",
+};
+
+/**
+ * If "mock" is given as a parameter, the function can return mocked data for a specific case. This
+ * pattern allows frontend developers to use your API before you finished implementing it!
+ *
+ * In addition, using Zod schemas to parse the response will make the input/output well-typed, making the code cleaner.
+ */
+export const chapterRequest = async (
```
Comment (nickbar01234) [inline src/app/api/chapter-request/route.client.ts:51]: nit: `createChapterRequest` is more intuitive.

File: src/app/api/chapter-request/route.spec.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+      universityEmail: "donaldduck@gmail.com",
+      phoneNumber: "1234567890",
+      university: "University of Waterloo",
+      universityAddress: "123 University Ave",
+      leadershipExperience: "I have lots of leadership experience",
+      motivation: "I am very motivated",
+      availabilities: "I am available",
+      questions: "I have a question",
+    };
+    const response = await fetch("http://localhost:3000/api/chapter-request", {
```
Comment (nickbar01234) [inline src/app/api/chapter-request/route.spec.ts:36]: Does this test run?

File: src/app/api/chapter-request/route.ts [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+          message: "Unknown error received",
+        }),
+        { status: 500 }
+      );
+    })
+  );
+  if (!chapterRequest.success) {
+    const error = chapterRequest.error.format();
+    // Check if any field fails the schema
+    if (Object.values(error).some((field) => !!field)) {
```
Comment (nickbar01234) [inline src/app/api/chapter-request/route.ts:25]: Can you elaborate on this logic? If `ChapterRequest` fails to parse, wouldn't this line always evaluate to true?

File: src/app/api/chapter-request/route.spec.ts (hunk 8 above)
Comment (cledi01) [reply src/app/api/chapter-request/route.spec.ts:36]: Yep

Comment (nickbar01234) [reply src/app/api/chapter-request/route.spec.ts:36]: Ah you have to start a server at the same time too right? I think we should avoid network calls in unit tests. An alternative option is exploring how to mock Prisma client to replicate "duplicate email". What do you think?

**Review by nickbar01234 — APPROVED** [review]
LGTM! The changes look amazing. Feel free to merge whenever :)

**Pull request merged**